### PR TITLE
WIP alternative overlay driver using multiple ro layers

### PR DESCRIPTION
**Pull request by vikstrous**
I started working on an overlay storage driver that doesn't use so many hardlinks and instead uses multiple read-only layers. The code is still pretty rough, but I wanted to open this PR to get some initial feedback on the idea.

I'm basically merging a bunch of code from the overlay driver and the aufs driver to produce an overlay driver that uses the directory structure and general approach of the aufs driver. Instead of hardlinking the lower layer into the upper layer when applying a diff, I'm just taking advantage of multiple read-only mounts. This feature was added to overlay in linux 3.19.

Major TODOs left:
- [x] convert overlayfs whiteouts and opaque directories to the AUFS format
- [ ] handle more than one page worth of mount arguments by chaining multiple mounts - I think we can simply Get an intermediate parent, then release them when we are released. We might have to store some metadata about parents in the `layers` file, indicating if the lowest parent is a `mnt` or `diff` .... maybe we shouldn't Get them because we really want those layers to be ro
- [x] detect support based on kernel version

I still have to clean up the code quite a bit, but it seems to work so far as long as  you don't have too many layers for one container and as long as you don't push/pull images that need to also work with other drivers.

I don't have any tests yet, so any tips for testing this would be appreciated.

I created this PR against 1.9 because it's stable, but there's no reason this wouldn't work with minimal modifications against master as well.

Other TODOs:
- [ ] Make sure it works with user namespaces
- [x] Add back the checks for incompatible underlying filesystem from the overlay driver
- [x] Return more info to docker inspect (GetMetadata)
- [ ] Think of more info to show in Status
- [x] Implement Changes correctly (what is it used for?) Should this Get the layer and its parent and do a Diff like NaiveDiffDriver or should this do the diff directly? 
- [ ] Do we need to do anything on Cleanup?
- [ ] Confirm that DiffSize doesn't need to be 100% exact (the size of xattrs vs opaque aufs files, etc. is different)
- [ ] Write tests

### CODE REVIEW

**Comment by tonistiigi** [comment]
Didn't look the code too much but couple of pointers:

Ideally, this should replace overlay and old driver would become overlay-legacy. Users with kernels less than 4.0 or with old graph directories would keep using the legacy driver while new and fresh kernels would start using this.

Whiteouts need to be replaced in both `Diff` and `ApplyDiff`. `Diff` changes to AUFS format and `ApplyDiff` back to overlay. If it helps then initially `ApplyDiff` could keep using `NaiveDiff` because this does not affect performance, but it's better to do the replacements because current implementation for `.wh.opq` in `ApplyUncompressedLayer` is not ideal.

We only accept pull requests against master branch atm.

File: daemon/graphdriver/overlay2/overlay2.go [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+}
+
+// Diff produces an archive of the changes between the specified
+// layer and its parent layer which may be "".
+func (d *Driver) Diff(id, parent string) (archive.Archive, error) {
+	// overlay2 doesn't need the parent layer to produce a diff.
+
+	// TODO: change format of whiteout files from char devices to the AUFS prefix?
+	// -> .wh..wh..opq .wh..wh
+	// XXX: if aufs excludes whiteout files, how can it restore the diff correctly???
```
Comment (tonistiigi) [inline daemon/graphdriver/overlay2/overlay2.go]: Only "meta" whiteouts are excluded, not file whiteouts or opaque directories.

Comment (vikstrous) [inline daemon/graphdriver/overlay2/overlay2.go]: Yep. I figured that out eventually. The comment is old.

**Comment by vikstrous** [comment]
I suspect that having multiple ro layers might have a worse performance at runtime than the hardlinks approach, so before making this the official overlay driver, I would like the benchmark the two.

**Comment by cpuguy83** [comment]
@vikstrous Can you sign your commit?
Also +1, thanks! I was looking at this as well.

**Comment by vikstrous** [comment]
I'm most of the way there, but there's some bug with the way I'm doing multi-level mounts that I can't figure out.

They seem to mostly work, but as soon as the daemon tries to start the container, something breaks and I get a useless error message.

I tried to add logs of log statements throughout my code and I got the following result: https://gist.github.com/vikstrous/2e79c66c2316eda763a0

The only information I have about what's failing is:
```
WARN[0093] exit status 1
```

It seems to happen after starting the container.

This is the dockerfile for the container I tried to build: https://gist.github.com/vikstrous/1ec40da0ab76f9f9e44f

Basically, it fails to start after layer 38.

Any tips for debugging this would be appreciated.

**Comment by tonistiigi** [comment]
@vikstrous so is the 38th layer the place where you hit the 4KB mount opt limit?

**Comment by vikstrous** [comment]
yep

**Comment by vikstrous** [comment]
I figured out the issue. If you set up mounts like this you can read, but not execute from the lower mount:

```
mount -t overlay overlay -olowerdir=lower:lower2 romerge
mount -t overlay overlay -olowerdir=romerge,upperdir=upper,workdir=work merged
./merged/ls
Failed to execute process './merged/ls'. Reason:
exec: No such device or address
```

I'm not sure if this is an overlay bug or it was just never meant to work this way. If I touch `ls` first, it copies it up and it works.

```
uname -a
Linux linusputinmusk 4.2.5-1-ARCH #1 SMP PREEMPT Tue Oct 27 08:13:28 CET 2015 x86_64 GNU/Linux
```

**Comment by vikstrous** [comment]
With the current behaviour of overlayfs, this approach is not going to work at all. I'll have to try it on the latest version of the kernel and if it fails there too, this PR is blocked on fixing overlayfs.

**Comment by vikstrous** [comment]
I opened an issue on the linux kernel bug tracker: https://bugzilla.kernel.org/show_bug.cgi?id=109601

**Comment by vikstrous** [comment]
I opened one more related issue: https://bugzilla.kernel.org/show_bug.cgi?id=109621
If we can remount with additional read-only layers, this will remove the need for a hierarchy of mounts and be much cleaner.

**Comment by calavera** [comment]
It looks like this change hit two issues in the kernel. Does it make sense to close it if we're not going to move it forward?

**Comment by vikstrous** [comment]
Considering that it's hopelessly blocked right now, yeah, let's just close it for now.

**Comment by vikstrous** [comment]
It looks like https://bugzilla.kernel.org/show_bug.cgi?id=109601 was fixed. This was the fix: https://github.com/torvalds/linux/commit/1c8a47df36d72ace8cf78eb6c228aa0f8027d3c2 It was a problem only between kernel versions 4.2 and 4.3 I'll make a new PR against master and we can probably get this into docker 1.11, but this new overlay driver will require kernel version 4.3.3.

**Comment by vikstrous** [comment]
Nevermind. This approach won't work. Overlays can be stacked only up to FILESYSTEM_MAX_STACK_DEPTH which is 2. This means the maximum number of layers is approximately 78 on docker 1.9 and will be smaller on docker 1.10 because of the new longer layer names.
